material-components/material-components-web
# chore(infrastructure): Retry failed GCP and CBT requests

### How to test

Unfortunately, this PR is hard to test, because it depends on the GCS/CBT servers returning HTTP errors and null response values.

Maybe some day we'll have unit tests for the screenshot testing infrastructure...

Until then, you kinda just have to rerun this thing a bunch of times until you see a "retry" message in the console output (see below for examples).

```bash
export CBT_USERNAME='...'
export CBT_AUTHKEY='...'
export MDC_GCLOUD_SERVICE_ACCOUNT_KEY_FILE_PATH='...'

# 0 diffs
npm run screenshot:build
npm run screenshot:test

# 2 diffs
echo '.mdc-button:not(:disabled){color:red}' >> packages/mdc-button/mdc-button.scss
echo '.mdc-button:disabled{margin-left:10px}' >> packages/mdc-button/mdc-button.scss
npm run screenshot:build
npm run screenshot:test
```

### What it does

- Retries failed GCS file upload requests due to:
    * HTTP server errors
- Retries failed CBT screenshot requests due to:
    * HTTP server errors
    * `null` screenshot image URLs returned by CBT
- Logs the queue index and queue size of file uploads

### Example output

#### Retrying failed GCP request

```
✔︎ Uploaded https://storage.googleapis.com/mdc-web-screenshot-tests/advorak/2018/05/18/22_36_41_063/7643f3bf1/lib/screenshot.js
➡ Uploading advorak/2018/05/18/22_36_41_063/7643f3bf1/out/mdc.floating-label.css.map ...
✗︎ FAILED to upload https://storage.googleapis.com/mdc-web-screenshot-tests/advorak/2018/05/18/22_36_41_063/7643f3bf1/lib/storage.js:
{ errors:
   [ { domain: 'global',
       reason: 'backendError',
       message: 'Backend Error' } ],
  code: 500,
  message: 'Backend Error' }
ERROR: GCP server returned HTTP 500. Retrying upload request...
✔︎ Uploaded https://storage.googleapis.com/mdc-web-screenshot-tests/advorak/2018/05/18/22_36_41_063/7643f3bf1/mdc-button/mixins/filled-accessible.html
✔︎ Uploaded https://storage.googleapis.com/mdc-web-screenshot-tests/advorak/2018/05/18/22_36_41_063/7643f3bf1/out/mdc.elevation.css
✗︎ FAILED to upload https://storage.googleapis.com/mdc-web-screenshot-tests/advorak/2018/05/18/22_36_41_063/7643f3bf1/mdc-button/mixins/icon-color.html:
{ errors:
   [ { domain: 'global',
       reason: 'backendError',
       message: 'Backend Error' } ],
  code: 500,
  message: 'Backend Error' }
ERROR: GCP server returned HTTP 500. Retrying upload request...
```

#### Logging file upload queue index and length

```
➡ Uploading file 110 of 112 - advorak/2018/05/18/22_46_52_895/7643f3bf1/test-cli.js ...
✔︎ Uploaded file 101 of 112 - https://storage.googleapis.com/mdc-web-screenshot-tests/advorak/2018/05/18/22_46_52_895/7643f3bf1/README.md
➡ Uploading file 78 of 112 - advorak/2018/05/18/22_46_52_895/7643f3bf1/out/mdc.ripple.css.map ...
✔︎ Uploaded file 98 of 112 - https://storage.googleapis.com/mdc-web-screenshot-tests/advorak/2018/05/18/22_46_52_895/7643f3bf1/out/mdc.top-app-bar.css.map
✔︎ Uploaded file 106 of 112 - https://storage.googleapis.com/mdc-web-screenshot-tests/advorak/2018/05/18/22_46_52_895/7643f3bf1/run-screenshot-test.js
✔︎ Uploaded file 107 of 112 - https://storage.googleapis.com/mdc-web-screenshot-tests/advorak/2018/05/18/22_46_52_895/7643f3bf1/run-static-server.js
```
